### DISCUSSION TITLE
How to make the right constraint settings

### DISCUSSION THREAD

**Discussion by yunjiandaiju** [yalmip/YALMIP#1353] [General discussion]
Here's my code.
`clc;
clear;
X=[1 0.5 -0.25;0 1 1];
U=[-1 -1];
X_0=[1 0.5;
     0  1];

X_1=[0.5 -0.25;
      1    1];
T1=inv(X_0)
X_1*T1
K1=U*T1
P = sdpvar(2,2,'full');
A=X_0*P;
B=X_1*P;
LMI1=[A  B
      B' A]
LMI2=[A-A']
Cons = [LMI1>=0,LMI2==0]
objective = [];
sol = optimize(Cons,objective);
P=double(P)
`
My LMI1 should be a matrix inequality constraint, why am I viewing it as an elemental constraint.
![QQ截图20231205145319](https://github.com/yalmip/YALMIP/assets/142418156/76953a1e-315b-48af-af67-33b51793f9a1)

**Comment by yunjiandaiju** [comment]
```
clc;
clear;
X=[1 0.5 -0.25;0 1 1];
U=[-1 -1];
X_0=[1 0.5;
     0  1];

X_1=[0.5 -0.25;
      1    1];
T1=inv(X_0)
X_1*T1
K1=U*T1
P = sdpvar(2,2,'full');
A=X_0*P;
B=X_1*P;
LMI1=[A  B
      B' A]
LMI2=[A-A']
Cons = [LMI1>=0,LMI2==0]
objective = [];
sol = optimize(Cons,objective);
P=double(P)

```

**Reply by johanlofberg** [reply]
This is just conceptually wrong. The matrix LMI1 is not structurally symmetric, hence the constraint LMI1>= 0 is not an LMI. It is also conceptually wrong to give the expression A-A' the name LMI2. LMI means linear matrix inequality, so to call it LMI2 just makes no sense. It signals that you don't understand what you are constructing.

My guess is that the underlying theory that you try to implement simply is incorrect, such as a flawed Lyapunov theory where you are missing symmetric parts of a model, such as a constraint AX+X'A' >= BB' which you incorrerctly have written as AX >=BB' or something like that, and then you've tried to implement that incorrect model

**Reply by yunjiandaiju** [reply]
![QQ截图20231205164255](https://github.com/yalmip/YALMIP/assets/142418156/50317a31-baaf-4a37-b21a-0cd2bfa96c70)
This is my constraint, did I define it wrong?

**Reply by johanlofberg** [reply]
You cannot model it like that. YALMIP will see a non-symmetric costraint in the second constraint and thus interpret as elementwise

However, I am 99% sure you simply try to implement the wrong model, i.e. the derivation of the thing that ends up in (20) is wrong as it almost never looks like that and X_theta would be symmetric by construction if theory done properly

**Reply by yunjiandaiju** [reply]
But the constraint considers X_*theta=(X_*theta)^T,which considers him symmetric, so the second constraint is also a symmetric matrix.

**Reply by yunjiandaiju** [reply]
I need to solve for a theta value, the first constraint requires X_*theta to be symmetric, the second is a symmetric constraint, how should I go about defining the constraints, X_ and X+ are known.

**Reply by johanlofberg** [reply]
You are confusing constraint and solution. The *constraint* is non-symmetric (and thus not interpreted as an LMI) but you have added an additional constraint asking for the *solution* to be symmetric. YALMIP analyzes the constraints, not the possible solutions (impossible, as that is for the solver to find)

[a b;b' a]>=0 will be elementwise if a is non-symmetric, and is still so even if you add a == a'. To fool YALMIP, you will have to introduce a new  symmetric decision variable z and use the model [z b;b' z]>=0, a == z

But as I said, I think your model is wrong, coming either from a missing symmetric part in some previous Lyapunov reasoning, erroneously trying to apply semidefinite programming to reason about eigenvalues of non-symmetric matrices, or some incorrect application of trying to linearize a quadratic semidefinite constraint. Impossible to say without you revealing where your constraints come from

**Reply by johanlofberg** [reply]
BTW regardless the model is ill-posed as P=0 would be feasible (if it was a correct model...).

**Comment by yunjiandaiju** [comment]
Okay, thank you very much for answering my question
